### PR TITLE
Compute lazy lambda serialization

### PR DESCRIPTION
**Pull request by RazvanN7**
This is an optimization to compute the lambda serialization only when __traits(isSame) is used on lambda functions.

### CODE REVIEW

File: src/dmd/lambdacomp.d [hunk 1]
```diff
@@ -262,9 +270,34 @@ public:
...
+        }
+        Dsymbol scopesym;
+        Dsymbol s = sc.search(t.sym.loc, t.sym.ident, &scopesym);
+        writeMangledName(s);
+    }
+
+    override void visit(TypeClass t)
     {
-        buf.reset();
+        if (LOG)
```
Comment (stefan-koch-sociomantic) [inline src/dmd/lambdacomp.d]: static if

File: src/dmd/lambdacomp.d [hunk 2]
```diff
@@ -83,6 +86,8 @@ public:
     override void visit(FuncLiteralDeclaration fld)
     {
         assert(fld.type.ty != Terror);
+        if (LOG)
```
Comment (stefan-koch-sociomantic) [inline src/dmd/lambdacomp.d]: static if

**Review by andralex — APPROVED** [review]
cc @WalterBright

**Comment by JinShil** [comment]
There aren't any tests and coverage is only 80%.  Is that deliberate?

**Comment by RazvanN7** [comment]
@JinShil This PR does not introduce any new functionality, but optimizes the lambda comparison. The code is not covered because the struct and class cases are identical and I only added examples for one of them. I will add another few tests so that coverage is 100%

**Comment by timotheecour** [comment]
@RazvanN7  speaking of __traits(isSame) and lambdas, there's a regression in git master:
https://issues.dlang.org/show_bug.cgi?id=18430 Issue 18430 - isSame is wrong for non global lambdas; obviously this PR couldn't have caused it since it's not merged but maybe you'd have some idea? (sorry for the drop in)

File: src/dmd/lambdacomp.d [hunk 3]
```diff
@@ -34,6 +35,8 @@ import dmd.statement;
 import dmd.tokens;
 import dmd.visitor;
 
+enum LOG = false;
```
Comment (wilzbach) [inline src/dmd/lambdacomp.d]: Out of interest: why not `version= LOG`? This is commonly used pattern in DRuntime.

File: src/dmd/traits.d [hunk 4]
```diff
@@ -1453,8 +1453,24 @@ extern (C++) Expression semanticTraits(TraitsExp e, Scope* sc)
 
         if (l1 && l2)
         {
-            import core.stdc.string;
+            const(char)* getSerialization(FuncLiteralDeclaration fld)
+            {
+                import dmd.lambdacomp;
```
Comment (wilzbach) [inline src/dmd/traits.d]: Nit: local imports should be selective.

File: src/dmd/traits.d [hunk 5]
```diff
@@ -1453,8 +1453,24 @@ extern (C++) Expression semanticTraits(TraitsExp e, Scope* sc)
 
         if (l1 && l2)
         {
-            import core.stdc.string;
+            const(char)* getSerialization(FuncLiteralDeclaration fld)
```
Comment (wilzbach) [inline src/dmd/traits.d]: Why don't you return `string` here? Comparing D strings is more efficient (no `\0` checking)
You already know the length of the buffer - no need to throw this valuable information away.

File: src/dmd/lambdacomp.d (hunk 3 above)
Comment (RazvanN7) [reply src/dmd/lambdacomp.d]: This is the pattern used in dmd from what I could obeserve :+1: 

https://github.com/dlang/dmd/blob/master/src/dmd/mtype.d#L61
https://github.com/dlang/dmd/blob/master/src/dmd/expression.d#L67
https://github.com/dlang/dmd/blob/master/src/dmd/expressionsem.d#L73

And the list could continue.

File: src/dmd/traits.d (hunk 5 above)
Comment (RazvanN7) [reply src/dmd/traits.d]: I know it's easier with strings, but the serialization field is is `const(char)*` and I really haven't seen AST members declared as strings. All of them are `const(char)*`

Comment (wilzbach) [reply src/dmd/traits.d]: Quoting a conversation with Walter here:

>> On 12/5/2017 11:20 AM, Andrei Alexandrescu wrote:
Hi Walter, Seb noticed that the dmd code is going through unpleasant hoops caused by C-style strings. Do you think incrementally switching to D-style strings would improve the compiler? -- Andrei

> Yes, but we have to be mindful of gdc and ldc both relying on the C++ interface to use dmd, and D strings don't work on that interface.
> I've already converted a bunch of internal stuff to D strings where the interface to the back end was not involved.

(we can even convert the C++ interfaces, but that's significantly more effort)

Comment (RazvanN7) [reply src/dmd/traits.d]: So how do we proceed here? That was my problem also: I could turn the field into a string, but what should I do regarding the header file?

**Comment by RazvanN7** [comment]
I discovered some problems. Please do not merge.

File: src/dmd/traits.d (hunk 5 above)
Comment (timotheecour) [reply src/dmd/traits.d]: better yet: for all the apis used by ldc, gdc
use a compiler dependent (via eg version(LDC)) type Dstring, which for dmd is alias to char[] and for ldc, dmd is alias to extern(C++) struct with size and char* head, along with conversions functions.
that way, dmd code stays clean, safe and keeps length aroud

Comment (timotheecour) [reply src/dmd/traits.d]: how about this:
```d
extern(C++) Dstring funLib(Dstring input);
```
see full working example here showing using such API from C++ with the function defined in D.

https://github.com/timotheecour/dtools/blob/master/tests/d01_dmd_string/

Comment (andralex) [reply src/dmd/traits.d]: @timotheecour #7893  is a good step in the right direction, we need to start using D strings in the compiler. That would be an effort separated from this.

**Review by andralex — APPROVED** [review]

**Pull request merged**